openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py

```diff
@@ -0,0 +1,444 @@
+                    print(f'Accuracy improved to {self.aggregated_model_accuracy} for round {self.round}')
+                    self.top_model_accuracy = self.aggregated_model_accuracy
+            
+            if self.dp_params is not None:
+                global_data_loader = DataLoader(self.collaborators, batch_size=int(self.sample_rate * float(len(self.collaborators))))            
+                dp_data_loader = DPDataLoader.from_data_loader(global_data_loader, distributed=False)
+                collaborator_batch = []
+                batch_sizes = []
+                for cols in dp_data_loader:
+                    batch_sizes.append(len(cols))
+                    collaborator_batch.append(cols)
+                self.round_collaborators = collaborator_batch[0]
+            else:
+                self.round_collaborators = self.collaborators                
+
+            self.round += 1
+            
+            if len(self.round_collaborators) !=0:
+                if not isinstance(self.round_collaborators[0], torch.Tensor):
+                    print(20*"#")
+                    print(f'Round {self.round}...')
+                    print("Batch sizes sampled:", batch_sizes)
+                    print(f"Collaborators patricipated in Round: {self.round}", self.round_collaborators)
+                    self.next(self.aggregated_model_validation, foreach='round_collaborators', exclude=['private'])
+                    print(20*"#")
+                else:
+                    print(f'No collaborator selected for training at Round: {self.round}')
+                    self.next(self.check_round_completion)
+            else:
+                print(f'No collaborator selected for training at Round: {self.round}')
+                self.next(self.check_round_completion)
+
+    @aggregator
+    def end(self):
+        print(20*"#")
+        print(f'All rounds completed successfully')
+        print(20*"#")
+        print(f'This is the end of the flow')
+        print(20*"#")
+
+if __name__ == '__main__':
+
+    argparser = argparse.ArgumentParser(
+        description=__doc__)
+    argparser.add_argument(
+        '--config_path',
+        help='Absolute path to the flow configuration file')
+    argparser.add_argument(
+        '--clip_test',
+        default=False, 
+        help='Indicate enabling of optimizer param testing before and after clip')
+
+    args = argparser.parse_args()
+
+    # Setup participants
+    aggregator = Aggregator()
+    aggregator.private_attributes = {}
+
+    # Setup collaborators with private attributes
+    collaborator_names = ['Portland', 'Seattle', 'Bangalore', 'Chandler', 'Austin', 
+                            'Hudson', 'Penang', 'Haifa', 'CostaRica', 'Guadalajara']
+    collaborators = [Collaborator(name=name) for name in collaborator_names]
+    
+    if torch.cuda.is_available():
+        device = torch.device(f'cuda:0')  # This will enable Ray library to reserve available GPU(s) for the task
```

Comment:
Devices should be maintained per collaborator. Right now, all processes are going to cuda 0.